### PR TITLE
[Resolve #1343]  Add feature to print out CloudFormation hook status to Sceptre output when logging new stack events

### LINKED ISSUE #1343: Ability to show the cloudformation hook status and message in the sceptre output
Cloudformation hook run during the cloudformation stack provisioning is not shown in sceptre output
We use the AWS cloudformation hooks for enforcing the compliance. But either in the Warn mode of Fail mode of the cloudformation hook - the sceptre output doesnot show the hook status, Hook status reason message.

Your environment
Sceptre version -- 4.2.0
python version -- 3.7.19**
which OS/distro -- Linux

Steps to reproduce
Tell us how to reproduce this issue. Please provide sceptre projct files if possible,
you can use https://plnkr.co/edit/ANFHm61Ilt4mQVgF as a base.

Expected behaviour
The sceptre output should show the Aws cloudformation "HookName", "HookStatus" and "HookStatusReason"

Actual behaviour
The sceptre output doesnot show the details of AWS Cloudformation hook.

(venv) [cloudshell-user@ip-10-4-13-6 cfn]$ sceptre create dev/sbucket.yaml
/home/cloudshell-user/venv/lib/python3.7/site-packages/sceptre/stack.py:430: DeprecatedWarning: role_arn is deprecated as of 4.0.0 and will be removed in 5.0.0. It is being renamed to "cloudformation_service_role". You should migrate all uses of "role_arn" to that in order to avoid future breakage.
setattr(self, deprecated_attribute_name, deprecated_value)
Do you want to create 'dev/sbucket.yaml' [y/N]: y
[2023-05-29 05:00:34] - dev/sbucket - Creating Stack
[2023-05-29 05:00:36] - dev/sbucket frp-dev-sbucket AWS::CloudFormation::Stack CREATE_IN_PROGRESS User Initiated
[2023-05-29 05:00:40] - dev/sbucket Bucket AWS::S3::Bucket CREATE_IN_PROGRESS
[2023-05-29 05:00:40] - dev/sbucket Bucket AWS::S3::Bucket CREATE_IN_PROGRESS
[2023-05-29 05:00:40] - dev/sbucket Bucket AWS::S3::Bucket CREATE_IN_PROGRESS
[2023-05-29 05:00:44] - dev/sbucket Bucket AWS::S3::Bucket CREATE_IN_PROGRESS
[2023-05-29 05:00:44] - dev/sbucket Bucket AWS::S3::Bucket CREATE_IN_PROGRESS
[2023-05-29 05:00:44] - dev/sbucket Bucket AWS::S3::Bucket CREATE_IN_PROGRESS Hook invocations complete. Resource creation initiated
[2023-05-29 05:00:44] - dev/sbucket Bucket AWS::S3::Bucket CREATE_IN_PROGRESS Resource creation Initiated
[2023-05-29 05:01:09] - dev/sbucket Bucket AWS::S3::Bucket CREATE_COMPLETE
[2023-05-29 05:01:09] - dev/sbucket frp-dev-sbucket AWS::CloudFormation::Stack CREATE_COMPLETE
(venv) [cloudshell-user@ip-10-4-13-6 cfn]$ sceptre --version
Sceptre, version 4.2.0
(venv) [cloudshell-user@ip-10-4-13-6 cfn]$ python3 --version
Python 3.7.16
(venv) [cloudshell-user@ip-10-4-13-6 cfn]$

### PR DESCRIPTION
**Pull request by fen9li**
[Resolve #1343] Add feature to print out CloudFormation hook status to Sceptre output when logging new stack events

    This pull request brings the ability for Sceptre to print out CloudFormation hook status when logging new stack events. They are "HookType","HookStatus","HookStatusReason" and "HookFailureMode". The information facilitates early picking up of CloudFormation hook issues/failures.

    Added CloudFormation hook information to `_log_new_events` method in `sceptre/plan/actions.py` module and its matching unit tests in `tests/test_actions.py` module.

    More details and discussion please refer [Ability to show the cloudformation hook status and message in the sceptre output #1343](https://github.com/Sceptre/sceptre/issues/1343) and [1343 logging hook status #1369](https://github.com/Sceptre/sceptre/pull/1369).

[1]:https://aws.amazon.com/about-aws/whats-new/2022/02/aws-announces-general-availability-aws-cloudformation-hooks/
[2]:https://docs.aws.amazon.com/AWSCloudFormation/latest/APIReference/API_StackEvent.html

### CODE REVIEW

File: sceptre/plan/actions.py [hunk 1]
```diff
@@ -848,6 +848,10 @@ def _log_new_events(self, after_datetime: datetime) -> datetime:
                         event["ResourceType"],
                         event["ResourceStatus"],
                         event.get("ResourceStatusReason", ""),
+                        event["HookType"],
+                        event["HookStatus"],
+                        event.get("HookStatusReason", ""),
+                        event["HookFailureMode"],
```
Comment (jfalkenstein) [inline sceptre/plan/actions.py:854]: I'd like to see examples of what the logs look like when there IS hook information and when there ISN'T hook information. Can you post some screenshots? I'm hesitant to add four new datapoints to every single log message if it's going to clutter the logs when 99% of events won't involve any hooks

File: tests/test_actions.py [hunk 2]
```diff
@@ -996,6 +996,10 @@ def test_log_new_events_prints_correct_event(self, mock_describe_events):
                     "LogicalResourceId": "id-2",
```
Comment (jfalkenstein) [inline tests/test_actions.py:996]: This test doesn't actually test that the new log events are logged. Right now, it's not really testing anything at all. You should update this test to actually test for the behavior that we expect log_new_events to have.

File: sceptre/plan/actions.py (hunk 1 above)
Comment (fen9li) [reply sceptre/plan/actions.py:854]: Hey @jfalkenstein 
Thank you for your time and effort. I tried to collect some screenshots for your reference. Please let me know if they help to make things clear. 

Please see below the No.1 screenshot. This is what I can see from Sceptre output, which was running in a Jenkins pipeline. 
test-cf-hook-sceptre-01

As I have configure the cloudformation hook mode to 'WARN', the stack was created without any blocking and I was thinking everythin works just fine.

However, if I double check the cloudformation events from aws web console for more details, I can see below:
test-cf-hook-web-console-06

That was saying the hook's lambda function does not work at all.

If Sceptre can display more details about cloudformation hooks, such as 'HookStatusReason', I can pick up this issue immediately from my Jenkins pipeline console. Otherwise, I would think my hook is doing fine until I dig deeper in aws web console.

In short, I hope Sceptre can support this feature in next release please.

Thank you.
Feng

File: tests/test_actions.py (hunk 2 above)
Comment (fen9li) [reply tests/test_actions.py:996]: Hey @jfalkenstein 
Thank you again for your time and effort.

Yeah, that was the first try run. I know I need to update the integration test code to accomodate my new 'log_new_events'. Was thinking there might be 1 or 2 failures and then I could fix them without too much efforts. Unfortunately, I saw almost half of the testing scenarios failed and was not quite sure where to start to update the integration tests code. Could you please shed some lights on this?

Thank you.
Feng

Comment (jfalkenstein) [reply tests/test_actions.py:996]: Yeah, you'll want to set your region to eu-west-1 when running integration tests. Otherwise, a number of them will fail.

Comment (fen9li) [reply tests/test_actions.py:996]: Hey @jfalkenstein 
Thank you for your response.

To try run the integration tests, I enter below command in the project root directory...

```
[fli@almalinux sceptre]$ pwd
/home/fli/github/sceptre
[fli@almalinux sceptre]$ 

[fli@almalinux sceptre]$ AWS_PROFILE=sceptre AWS_DEFAULT_REGION=eu-west-1 poetry run behave --junit --junit-directory build/behave integration-tests
...
[fli@almalinux sceptre]$ 
```

The integration test reports were collected successfully ...
```
[fli@almalinux sceptre]$ ls build/behave/
TESTS-features.create-change-set.xml               TESTS-sceptre.integration-tests.features.create-change-set.xml
TESTS-features.create-stack.xml                    TESTS-sceptre.integration-tests.features.create-stack.xml
TESTS-features.delete-change-set.xml               TESTS-sceptre.integration-tests.features.delete-change-set.xml
TESTS-features.delete-stack-group.xml              TESTS-sceptre.integration-tests.features.delete-stack-group.xml
TESTS-features.delete-stack.xml                    TESTS-sceptre.integration-tests.features.delete-stack.xml
TESTS-features.dependency-resolution.xml           TESTS-sceptre.integration-tests.features.dependency-resolution.xml
TESTS-features.describe-change-set.xml             TESTS-sceptre.integration-tests.features.describe-change-set.xml
TESTS-features.describe-stack-group-resources.xml  TESTS-sceptre.integration-tests.features.describe-stack-group-resources.xml
TESTS-features.describe-stack-group.xml            TESTS-sceptre.integration-tests.features.describe-stack-group.xml
TESTS-features.describe-stack-resources.xml        TESTS-sceptre.integration-tests.features.describe-stack-resources.xml
TESTS-features.drift.xml                           TESTS-sceptre.integration-tests.features.drift.xml
TESTS-features.dump-template-s3.xml                TESTS-sceptre.integration-tests.features.dump-template-s3.xml
TESTS-features.dump-template.xml                   TESTS-sceptre.integration-tests.features.dump-template.xml
TESTS-features.execute-change-set.xml              TESTS-sceptre.integration-tests.features.execute-change-set.xml
TESTS-features.launch-stack-group.xml              TESTS-sceptre.integration-tests.features.launch-stack-group.xml
TESTS-features.launch-stack.xml                    TESTS-sceptre.integration-tests.features.launch-stack.xml
TESTS-features.list-change-sets.xml                TESTS-sceptre.integration-tests.features.list-change-sets.xml
TESTS-features.lock-stack.xml                      TESTS-sceptre.integration-tests.features.lock-stack.xml
TESTS-features.project-dependencies.xml            TESTS-sceptre.integration-tests.features.project-dependencies.xml
TESTS-features.prune.xml                           TESTS-sceptre.integration-tests.features.prune.xml
TESTS-features.stack-diff.xml                      TESTS-sceptre.integration-tests.features.stack-diff.xml
TESTS-features.stack-output-external-resolver.xml  TESTS-sceptre.integration-tests.features.stack-output-external-resolver.xml
TESTS-features.stack-output-resolver.xml           TESTS-sceptre.integration-tests.features.stack-output-resolver.xml
TESTS-features.unlock-stack.xml                    TESTS-sceptre.integration-tests.features.unlock-stack.xml
TESTS-features.update-stack.xml                    TESTS-sceptre.integration-tests.features.update-stack.xml
TESTS-features.validate-template.xml               TESTS-sceptre.integration-tests.features.validate-template.xml
[fli@almalinux sceptre]$ 
```

A typical passed report:

[TESTS-features.create-change-set.xml.log](https://github.com/Sceptre/sceptre/files/12297310/TESTS-features.create-change-set.xml.log)

A typical failed report:

[TESTS-features.create-stack.xml.log](https://github.com/Sceptre/sceptre/files/12297312/TESTS-features.create-stack.xml.log)

I reckon I need to update the integration code somewhere, but not very sure. Could you please advise ?

Thank you.
Feng

Comment (jfalkenstein) [reply tests/test_actions.py:996]: According to those failure logs you linked to... it looks like HookType isn't always returned in the cloudformation events. I don't think you need to fix the integration test code. I think you need to fix the event code to handle cases where there's no HookType key in the results.

Comment (fen9li) [reply tests/test_actions.py:996]: Hey @jfalkenstein 

Thank you for your help. I am editing my change as below:

```
PS C:\Users\fli\github\sceptre> git diff .\sceptre\plan\actions.py
diff --git a/sceptre/plan/actions.py b/sceptre/plan/actions.py
index bebd09c..78d599d 100644
--- a/sceptre/plan/actions.py
+++ b/sceptre/plan/actions.py
@@ -840,21 +840,34 @@ class StackActions:
         events.reverse()
         new_events = [event for event in events if event["Timestamp"] > after_datetime]
         for event in new_events:
-            self.logger.info(
-                " ".join(
-                    [
-                        self.stack.name,
-                        event["LogicalResourceId"],
-                        event["ResourceType"],
-                        event["ResourceStatus"],
-                        event.get("ResourceStatusReason", ""),
-                        event["HookType"],
-                        event["HookStatus"],
-                        event.get("HookStatusReason", ""),
-                        event["HookFailureMode"],
-                    ]
+            if "HookStatus" in event:
+                self.logger.info(
+                    " ".join(
+                        [
+                            self.stack.name,
+                            event["LogicalResourceId"],
+                            event["ResourceType"],
+                            event["ResourceStatus"],
+                            event.get("ResourceStatusReason", ""),
+                            event["HookType"],
+                            event["HookStatus"],
+                            event.get("HookStatusReason", ""),
+                            event["HookFailureMode"],
+                        ]
+                    )
+                )
+            else:
+                self.logger.info(
+                    " ".join(
+                        [
+                            self.stack.name,
+                            event["LogicalResourceId"],
+                            event["ResourceType"],
+                            event["ResourceStatus"],
+                            event.get("ResourceStatusReason", ""),
+                        ]
+                    )
                 )
-            )
             after_datetime = event["Timestamp"]
         return after_datetime

PS C:\Users\fli\github\sceptre>
```

Will see how it is going. Cheers!

Feng

**Comment by fen9li** [comment]
Hey @jfalkenstein 

I hope you have a good day.
Thank you for your help. The integegration all passed successfully and the test reports have collected for further review. Please see below screenshot for reference.

sceptre-integration-test-01

sceptre-integration-test-02

Could you please advise anything else I need to do for this pull request? Cheers!

Feng

**Review by jfalkenstein — CHANGES_REQUESTED** [review]
I like how you fixed your code so it only adds those fields when they're present.

With that said, I'd like to see at least 2 unit tests over this code change: one test to cover cases where the fields are not present and one test to cover cases where the fields _are_ present. Both should test that the expected values are logged. 

If you're not familiar with how to capture logs in tests, you can take a look here: https://docs.pytest.org/en/7.1.x/how-to/logging.html#caplog-fixture.

**Comment by fen9li** [comment]
Thanks @jfalkenstein , will do as soon as I can. 
Cheers.
Feng

**Comment by fen9li** [comment]
Hey @jfalkenstein Thank you for your patience. I have added capture logs tests in the unit tests, for both with CF hooks and without CF hooks. Do you mind reviewing the pull request again when you have some time? Please let me know if you have more requests. Cheers!
Feng

**Comment by jfalkenstein** [comment]
This is great! I'm going to run integration tests and, assuming they all pass, we can merge.

**Comment by jfalkenstein** [comment]
Meanwhile, can you update the PR description to include a full description of your changes? (We use that for the full commit message body).

**Comment by fen9li** [comment]
@jfalkenstein Thank you for reviewing. Done.
Feng

**Comment by jfalkenstein** [comment]
@fen9li Can you please change the PR title to start with `[Resolve #1343]`? I can't update your title.

**Review by jfalkenstein — APPROVED** [review]
